src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -859,88 +847,113 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
   }
 }
 
-#define INSN(NAME, REGISTER)                                       \
-  void MacroAssembler::NAME(const address dest, Register temp) {   \
-    assert_cond(dest != nullptr);                                  \
-    int64_t distance = dest - pc();                                \
-    if (is_simm21(distance) && ((distance % 2) == 0)) {            \
-      Assembler::jal(REGISTER, distance);                          \
-    } else {                                                       \
-      assert(temp != noreg, "expecting a register");               \
-      int32_t offset = 0;                                          \
-      movptr(temp, dest, offset);                                  \
-      Assembler::jalr(REGISTER, temp, offset);                     \
-    }                                                              \
-  }                                                                \
-
-  INSN(j,   x0);
-  INSN(jal, x1);
-
-#undef INSN
+void MacroAssembler::jump_link(const address dest, Register temp) {
+  assert_cond(dest != nullptr);
+  int64_t distance = dest - pc();
+  if (is_simm21(distance) && ((distance % 2) == 0)) {
+    Assembler::jal(x1, distance);
+  } else {
+    assert(temp != noreg && temp != x0, "expecting a register");
+    int32_t offset = 0;
+    movptr(temp, dest, offset);
```

Comment:
One more question: will `la(temp, dest, offset)` work here? I see you've changed to use `la` for `MacroAssembler::j(const address dest, Register temp)` instead of the `movptr` used before Seems we can do similar change here as well.

Reply:
Jump can 'only' go to code cache (if someone jumps to C/C++ method we would be in trouble).
Therefore I changed to la(), otherwise I tried to keep generate code same:ish.

I don't see any reason for why we can't use it for those cases.

I think any movptr in a compressible region should be la().

Sanity tested with RCC 2047 and unset.